plugin-viewport/src/main/kotlin/com/mapbox/maps/plugin/viewport/experimental/ViewportPluginImpl.kt

```diff
@@ -0,0 +1,367 @@
+      type: CameraAnimatorType,
+      runningAnimator: ValueAnimator,
+      runningAnimatorOwner: String?,
+      newAnimator: ValueAnimator,
+      newAnimatorOwner: String?
+    ) {
+      // no-ops
+    }
+
+    override fun onAnimatorEnding(
+      type: CameraAnimatorType,
+      animator: ValueAnimator,
+      owner: String?
+    ) {
+      // no-ops
+    }
+
+    override fun onAnimatorCancelling(
+      type: CameraAnimatorType,
+      animator: ValueAnimator,
+      owner: String?
+    ) {
+      // no-ops
+    }
+  }
+
+  /**
+   * Returns list of registered states.
+   */
+  override val states: Set<ViewportState>
+    get() = internalStates
+
+  /**
+   * Returns current [ViewportStatus].
+   *
+   * If current status is IDLE, returns ViewportStatus.State(null).
+   * @see addStatusObserver
+   */
+  override var status: ViewportStatus = ViewportStatus.State(null)
+    @Synchronized private set
+    @Synchronized get
+
+  private fun updateStatus(targetStatus: ViewportStatus, reason: String) {
+    val previousStatus = status
+    status = targetStatus
+    notifyStatusChanged(previousStatus, targetStatus, reason)
+  }
+
+  /**
+   * Executes a transition to requested state.
+   *
+   * When started, goes to [ViewportTransition]
+   * and to the final [ViewportState] when ended.
+   *
+   * If transition is canceled, state goes to IDLE.
+   *
+   * @param targetState The target [ViewportState] to transition to.
+   * @param completionListener The listener to observe the completion state.
+   */
+  override fun transitionTo(targetState: ViewportState, completionListener: CompletionListener?) {
+    with(status) {
+      when (this) {
+        is ViewportStatus.State -> {
+          if (this.state === targetState) {
+            completionListener?.onComplete(true)
+            return
+          }
+        }
+        is ViewportStatus.Transition -> {
+          if (this.toState === targetState) {
+            completionListener?.onComplete(false)
+            return
+          }
+        }
+      }
+    }
+    val fromState = status.getCurrentOrTargetState()
+    currentCancelable?.cancel()
+    currentCancelable = null
+    // implicit adding of states upon transitioning to a state that hasn't been added yet
+    if (!states.contains(targetState)) {
+      internalStates.add(targetState)
+    }
+
+    // get the transition (or default) for the from and to state
+    val transition = getTransition(fromState, targetState) ?: defaultTransition
+
+    // run the transition
+    var completionBlockInvoked = false
+    val transitionCancelable = transition.run(fromState, targetState) { isFinished ->
+      completionBlockInvoked = true
+      if (isFinished) {
+        // transfer camera updating responsibility to targetState
+        currentCancelable = targetState.startUpdatingCamera()
+        updateStatus(
+          ViewportStatus.State(targetState),
+          VIEWPORT_STATUS_OBSERVER_REASON_PROGRAMMATIC
+        )
```

Comment:
How should we handle the following:

* Several listeners are registered
* Transition A completes
* The first listener is notified and initiates Transition B
* The observers are notified about Transition B
* The remaining observers are notified about Transition A

You could make this scenario arbitrarily complex by adding more branches.

cc: @pengdev

Reply:
One option is to just disallow it. We could even assert by setting an internal flag while notifying. If we do this, we'd want to make sure that we notify *before* invoking completion blocks since completion blocks could be used to trigger the same behavior.

Another option would be to just skip delivering the rest of the Transition A notifications to avoid delivering them out of order. The downside is that observers would miss certain updates.

I'm leaning toward the first option, but I'll wait for your input.

Reply:
@macdrevx could you describe a bit more on the first option? does it mean that user can not transition to new state in the status observer?

Reply:
Yes, that's right. Attempting to start a new transition within a status change listener would assert.

Another option to consider is always delivering notifications async with a serial queue. This would mean we couldn't guarantee that toState in the notification matches the viewport's current state.

Reply:
went forward with delivering notifications to a message queue

Reply:
Same on iOS